File: src/typescript/src/transport/i-transport.ts
import { MessageInfo } from './transaction/message/message-info';
import { Transaction } from './transaction/transaction';

/**
 * An interface for a transport mechanism to send and receive messages.
 */
export interface ITransport {

  /**
   * Where messages that cannot be delivered are sent.
   */
  deadLetterEndpoint: string;

  /**
   * If the transport supports sending command messages, this will be true.
   */
  supportsCommandMessages: boolean;

  /**
   * If the transport supports sending delayed commands, this will be true.
   */
  supportsDelayedCommands: boolean;

  /**
   * If the transport supports event message subscriptions, this will be true.
   */
  supportsSubscriptions: boolean;

  /**
   * Sends messages associated with the given transaction to the transport.
   */
  // eslint-disable-next-line no-unused-vars
  handle(transaction: Transaction): Promise<void>;

  /**
   * Subscribes to a messages so that the transport can start receiving them.
   */
  // eslint-disable-next-line no-unused-vars
  subscribe(messageInfo: MessageInfo): Promise<void>;

  /**
   * Starts the transport to start processing messages.
   */
  start(): Promise<void>;

  /**
   * Stops the transport from processing messages.
   */
  stop(): Promise<void>;
}

File: src/python/src/transport/i_transport.py
"""Transport interface for the PolyBus Python implementation."""

from abc import ABC, abstractmethod
from src.transport.transaction.transaction import Transaction
from src.transport.transaction.message.message_info import MessageInfo


class ITransport(ABC):
    """An interface for a transport mechanism to send and receive messages."""
    
    @property
    @abstractmethod
    def dead_letter_endpoint(self) -> str:
        """Where messages that cannot be delivered are sent."""
        pass
    
    @abstractmethod
    async def handle(self, transaction: 'Transaction') -> None:
        """Sends messages associated with the given transaction to the transport.
        
        Args:
            transaction: The transaction containing messages to send.
        """
        pass
    
    @property
    @abstractmethod
    def supports_delayed_commands(self) -> bool:
        """If the transport supports sending delayed commands, this will be true."""
        pass
    
    @property
    @abstractmethod
    def supports_command_messages(self) -> bool:
        """If the transport supports sending command messages, this will be true."""
        pass
    
    @abstractmethod
    async def subscribe(self, message_info: 'MessageInfo') -> None:
        """Subscribes to a messages so that the transport can start receiving them.
        
        Args:
            message_info: Information about the message type to subscribe to.
        """
        pass
    
    @property
    @abstractmethod
    def supports_subscriptions(self) -> bool:
        """If the transport supports event message subscriptions, this will be true."""
        pass
    
    @abstractmethod
    async def start(self) -> None:
        """Starts the transport to start processing messages."""
        pass
    
    @abstractmethod
    async def stop(self) -> None:
        """Stops the transport from processing messages."""
        pass